File: app/Models/StudyGroupMember.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class StudyGroupMember extends Model
{
    use HasFactory;

    public const ROLE_ADMIN = 'admin';

    public const ROLE_MEMBER = 'member';

    public const ROLE_MODERATOR = 'moderator';

    public const ROLE_OWNER = 'owner';

    protected $fillable = [
        'study_group_id',
        'user_id',
        'role',
        'joined_at',
    ];

    protected function casts(): array
    {
        return [
            'joined_at' => 'datetime',
        ];
    }

    public function group(): BelongsTo
    {
        return $this->belongsTo(StudyGroup::class, 'study_group_id');
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function canManageFocusRooms(): bool
    {
        return in_array($this->role, [self::ROLE_OWNER, self::ROLE_ADMIN], true);
    }
}

File: app/models/study_group_member.rb
class StudyGroupMember < ApplicationRecord
  ROLE_ADMIN = "admin"
  ROLE_MEMBER = "member"
  ROLE_MODERATOR = "moderator"
  ROLE_OWNER = "owner"

  belongs_to :study_group
  belongs_to :user

  validates :role, presence: true

  # Only owners and admins may manage focus rooms inside a group.
  def can_manage_focus_rooms?
    [ ROLE_OWNER, ROLE_ADMIN ].include?(role)
  end
end
